lib/fs.js

```diff
@@ -1836,22 +1836,36 @@ fs.FileWriteStream = fs.WriteStream; // support the legacy name
 
 
 WriteStream.prototype.open = function() {
+  const self = this;
+
   fs.open(this.path, this.flags, this.mode, function(er, fd) {
     if (er) {
-      this.destroy();
-      this.emit('error', er);
+      self.destroy();
+      self.emit('error', er);
       return;
     }
 
-    this.fd = fd;
-    this.emit('open', fd);
-  }.bind(this));
+    self.fd = fd;
+    self.emit('open', fd);
+  });
 };
 
 
+function write(fd, buffer, position, callback) {
+  function wrapper(err, written) {
+    // Retain a reference to buffer so that it can't be GC'ed too soon.
+    callback(err, written || 0, buffer);
+  }
```

Comment:
You won't see much if any gain in terms of performance. The change is more for code cleanliness. Will have to make the change of passing `buffer` to `wrapper()` so it can be hoisted. Though that shouldn't be difficult.